tik/lib/Util.cpp

```diff
@@ -449,4 +486,68 @@ namespace TraceAtlas::tik
         }
         return result;
     }
-} // namespace TraceAtlas::tik
+
+    set<BasicBlock *> GetExits(set<BasicBlock *> blocks)
+    {
+        set<BasicBlock *> Exits;
+        for (auto block : blocks)
+        {
+            for (auto suc : successors(block))
+            {
+                if (blocks.find(suc) == blocks.end())
+                {
+                    Exits.insert(suc);
+                }
+            }
+            auto term = block->getTerminator();
+            if (auto retInst = dyn_cast<ReturnInst>(term))
+            {
+                Function *base = block->getParent();
+                for (auto user : base->users())
+                {
+                    if (auto v = dyn_cast<Instruction>(user))
+                    {
+                        auto parentBlock = v->getParent();
+                        if (blocks.find(parentBlock) == blocks.end())
+                        {
+                            Exits.insert(parentBlock);
+                        }
+                    }
+                }
+            }
+        }
+        return Exits;
+    }
+
+    set<BasicBlock *> GetExits(set<BasicBlock *> &s, BasicBlock *entrance)
+    {
+        set<BasicBlock *> result;
+        for (auto block : s)
+        {
+            for (auto suc : successors(block))
+            {
+                if (s.find(suc) == s.end() && suc->getParent() == entrance->getParent())
+                {
+                    result.insert(suc);
+                }
+            }
+            auto term = block->getTerminator();
+            if (auto retInst = dyn_cast<ReturnInst>(term))
+            {
+                Function *base = block->getParent();
+                for (auto user : base->users())
+                {
+                    if (auto v = dyn_cast<Instruction>(user))
+                    {
+                        auto parentBlock = v->getParent();
+                        if (s.find(parentBlock) == s.end() && parentBlock->getParent() == entrance->getParent())
+                        {
+                            result.insert(parentBlock);
+                        }
+                    }
+                }
+            }
+        }
+        return result;
+    }
```

Comment:
Did you rewrite the exit code? github marks the entire area as a delta.

If so, this code has been migrated to AtlasUtil.

Reply:
I added a function that is used when I collect function args in GetBounaryValues(). It creates an estimation for all exit blocks before we have the kernel function. Its purpose is to provide a set of blocks whose users may be outside the kernel via a premature exit

Reply:
Please check AtlasUtil and get back to me. An estimate should not be necessary. Is there a reason you can't use that code?

Reply:
Before we acquire the exits of the kernel (because our current paradigm requires the kernel function to be constructed first), the function args have to be found. When a kernel can exit early, and the predecessors to that exit hold values that are used later in the kernel (and thus, after an early exit, there will be users of this value in the original bitcode), the value has to be exported or else it is unresolved at compile time